### PR TITLE
Crusher for the new map

### PR DESCRIPTION
**Pull request by Arrow768**
This adds the crusher for the new map.

When mapping it in:
Add the /obj/machinery/crusher_base next to each other
Add a /obj/item/modular_computer/telescreen/preset/trashcompactor near the pistons
Change id_tag of the airlock leading into the crusher to "crusher"

The crusher can be built by using a crusher circuitboard.

ToDo to merge this pull:
- [x] Add status lights to the crusher base
- [x] Add more icon states for retracting the crusher from different states (so abort works correctly)
- [x] Improve the Control App
- [x] Add a check if the crusher piston can extend (check if theres machinery, walls, ... in the way)
- [x] Destruction and Deconstruction of the crusher

Testing ToDos
- [ ] Check if the mapped-in crusher works
- [ ] Build the crusher manually and check if it works
- [ ] Research the board (check if the sorting is fine)

### CODE REVIEW

File: code/game/machinery/crusher_piston.dm [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+	component_parts += new /obj/item/weapon/circuitboard/crusher(src)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin(src)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin(src)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin(src)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin(src)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(src)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(src)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(src)
+	component_parts += new /obj/item/weapon/reagent_containers/glass/beaker(src)
+	component_parts += new /obj/item/weapon/reagent_containers/glass/beaker(src)
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: Why does the crusher need beakers?

File: code/game/machinery/crusher_piston.dm [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+	component_parts += new /obj/item/weapon/reagent_containers/glass/beaker(src)
+	RefreshParts()
+
+	action_start_time = world.time
+
+	//Spawn the stage 1 pistons south of it with a density of 0
+	log_debug("Spawning Crusher Piston Stage")
+	pstn = new(locate(x,y-1,z))
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: ...what?

Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: This could probably be `get_step()`.

Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: This could probably be `get_step(src, SOUTH)`.

File: code/game/machinery/crusher_piston.dm [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+/obj/machinery/crusher_base/proc/get_status()
+	return status
+
+/obj/machinery/crusher_base/proc/is_blocked()
+	return blocked
+
+/obj/machinery/crusher_base/proc/valve_check() //Check if the depreasurization valve is open
+	if(valve_open == 1)
+		visible_message("The hydraulic pump in [src] briefly spins up and then shuts down","You hear a pump spinning up briefly and then shutting down")
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: Periods please.

File: code/game/machinery/crusher_piston.dm [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+	return ..()
+
+/obj/machinery/crusher_piston/proc/extend_0_1()
+	var/turf/T = get_turf(src)
+	if(!can_extend_into(T))
+		log_debug("cant extend 0-1 - Abort")
+		return 0
+	icon_state="piston_0_1"
+	stage = 1
+	pb1 = new(locate(x,y,z))
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: ...why not `new(loc)`?

File: code/game/machinery/crusher_piston.dm [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+/obj/machinery/crusher_piston/proc/extend_1_2()
+	var/turf/T = get_turf(pb1)
+	var/turf/extension_turf = get_step(T,SOUTH)
+	if(!can_extend_into(extension_turf))
+		log_debug("cant extend 1-2 - Abort")
+		return 0
+	icon_state="piston_1_2"
+	stage = 2
+	pb2 = new(locate(x,y-1,z))
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: `get_step()`.

File: code/game/machinery/crusher_piston.dm [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+	for(var/atom/movable/AM in get_turf(locate(x,y-1,z)))
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: `get_step()`.

File: code/game/machinery/crusher_piston.dm [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+// Return 1 if it could successfully be moved
+// Return 0 if it could not be moved
+/atom/movable/proc/piston_move()
+	var/turf/T = get_turf(src)
+
+	var/list/valid_turfs = list()
+	for(var/dir_to_test in cardinal)
+		var/turf/new_turf = get_step(T, dir_to_test)
+		if(!new_turf.contains_dense_objects())
+			valid_turfs |= new_turf
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: Might as well use `+=` here, no reason to use `|=`.

File: code/game/machinery/crusher_piston.dm [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+
+/mob/living/piston_move()
+	for(var/i = 1, i <= PISTON_MOVE_DIVISOR, i++)
+		adjustBruteLoss(round(PISTON_MOVE_DAMAGE / PISTON_MOVE_DIVISOR))
+	AdjustStunned(5)
+	AdjustWeakened(5)
+	return ..()
+
+/mob/living/carbon/piston_move()
+	if(!(species && (species.flags & NO_PAIN)))
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: Don't only `/mob/living/carbon/human` have a species?

File: code/game/machinery/crusher_piston.dm [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+	var/list/unmovable_items = list(
+		/obj/machinery,
+		/obj/structure,
+		/obj/item/modular_computer/telescreen,
+		/obj/item/modular_computer/console
+	)
+	if (istype(extension_turf,/turf/simulated/wall))
+		return 0
+	for(var/atom/A in extension_turf)
+		if (is_type_in_list(A, unmovable_items))
```
Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: Note to self: Typecache.

File: code/game/machinery/crusher_piston.dm (hunk 1 above)
Comment (Arrow768) [inline code/game/machinery/crusher_piston.dm]: To store the oil of the pistons

File: code/game/machinery/crusher_piston.dm (hunk 8 above)
Comment (Arrow768) [inline code/game/machinery/crusher_piston.dm]: grafik

Comment (Lohikar) [inline code/game/machinery/crusher_piston.dm]: TIL. No complaint then.

**Pull request merged**